p2p/module.go

```diff
@@ -177,15 +230,181 @@ func (m *p2pModule) Send(addr cryptoPocket.Address, msg *anypb.Any) error {
+	}
+}
+
+func (m *p2pModule) startNetwork() (err error) {
+	if m.cfg.UseRainTree {
+		m.network = raintree.NewRainTreeNetwork(
+			m.address,
+			m.GetBus(),
+			m.pstoreProvider,
+			m.currentHeightProvider,
+		)
+	} else {
+		m.network, err = network.NewLibp2pNetwork(m.GetBus(), m.logger, m.host, m.topic)
+	}
+	return err
+}
+
+func (m *p2pModule) startHost() (err error) {
+	// Return early if host has already been started (e.g. via `WithHostOption`)
+	if m.host != nil {
+		return nil
+	}
+	opts := []libp2p.Option{
+		// Explicitly specify supported transport security options (noise, TLS)
+		// (see: https://pkg.go.dev/github.com/libp2p/go-libp2p@v0.26.3#DefaultSecurity)
+		libp2p.DefaultSecurity,
+		m.identity,
+	}
+
+	// Disable unused libp2p relay and ping services in client debug mode.
+	// (see: https://pkg.go.dev/github.com/libp2p/go-libp2p#DisableRelay
+	// and https://pkg.go.dev/github.com/libp2p/go-libp2p#Ping)
+	if m.isClientDebugMode() {
+		opts = append(opts,
+			libp2p.DisableRelay(),
+			libp2p.Ping(false),
+			libp2p.NoListenAddrs,
+		)
+	} else {
+		opts = append(opts, m.listenAddrs)
+	}
+
+	m.host, err = libp2p.New(opts...)
+	if err != nil {
+		return fmt.Errorf("unable to create libp2p host: %w", err)
+	}
+	return nil
+}
+
+func (m *p2pModule) isClientDebugMode() bool {
+	return m.GetBus().GetRuntimeMgr().GetConfig().ClientDebugMode
+}
+
+// handleStream is called each time a peer establishes a new stream with this
+// module's libp2p `host.Host`.
+func (m *p2pModule) handleStream(stream libp2pNetwork.Stream) {
+	peer, err := utils.PeerFromLibp2pStream(stream)
+	if err != nil {
+		m.logger.Error().Err(err).
+			Str("address", peer.GetAddress().String()).
+			Msg("parsing remote peer public key")
+
+		if err = stream.Reset(); err != nil {
+			m.logger.Error().Err(err).Msg("resetting stream")
+		}
+	}
+
+	if err := m.network.AddPeer(peer); err != nil {
+		m.logger.Error().Err(err).
+			Str("address", peer.GetAddress().String()).
+			Msg("adding remote peer to address book")
+	}
+
+	go m.readStream(stream)
+}
+
+// readStream is intended to be called in a goroutine. It continuously reads from
+// the given stream for handling at the network level. Used for handling "direct"
+// messages (i.e. one specific target node).
+func (m *p2pModule) readStream(stream libp2pNetwork.Stream) {
+	// Time out if no data is sent to free resources.
+	if err := stream.SetReadDeadline(newReadStreamDeadline()); err != nil {
+		// NB: tests using libp2p's `mocknet` rely on this not returning an error.
+		// `SetReadDeadline` not supported by `mocknet` streams.
+		m.logger.Debug().Err(err).Msg("setting stream read deadline")
+	}
+
+	data, err := io.ReadAll(stream)
+	if err != nil {
+		m.logger.Error().Err(err).Msg("reading from stream")
+		if err := stream.Reset(); err != nil {
+			m.logger.Debug().Err(err).Msg("resetting stream")
+		}
+		return
+	}
+
+	m.handleNetworkData(data)
```

Comment:
Thanks for calling this out! :raised_hands: Since `#handleNetworkData()` is only being called by `#readStream()`, which is run in a go routine per stream, I hadn't noticed yet. I agree that it would be better to have it return an error. I think this moves all the error handling to one place, `readStream`, which is an improvement in maintainability and readability. :+1: